Extension/src/SSH/commandInteractors.ts

```diff
@@ -0,0 +1,369 @@
+
+        return result;
+    }
+}
+
+export class DifferingHostKeyInteractor implements IInteractor {
+    static ID = 'differing host key';
+
+    constructor(
+        private readonly confirmationProvider: IDifferingHostConfirmationProvider
+    ) { }
+
+    get id(): string {
+        return DifferingHostKeyInteractor.ID;
+    }
+    async onData(
+        data: string,
+        cancelToken?: vscode.CancellationToken,
+        _extraDetails?: IInteractorDataDetails
+    ): Promise<IInteraction> {
+        const result: IInteraction = { postAction: 'keep' };
+        data = data.trim();
+
+        if (
+            data.includes('Are you sure you want to continue connecting') &&
+            data.includes('Offending key for IP in') &&
+            data.includes('Matching host key in')
+        ) {
+            result.postAction = 'consume';
+            const message: string = data.substring(data.indexOf('Warning'), data.indexOf('Are')).trim();
+            const confirmation: string | undefined = await this.confirmationProvider(message, cancelToken);
+            if (confirmation) {
+                result.response = confirmation;
+            } else {
+                result.canceled = true;
+            }
+        }
+
+        return result;
+    }
+}
+
+export type IStringProvider =
+    (key?: string, detail?: string, cancelToken?: vscode.CancellationToken) => Promise<string | undefined>;
+
+export class PassphraseInteractor implements IInteractor {
+    static ID = 'passphrase';
+
+    constructor(private readonly passphraseProvider: IStringProvider) { }
+
+    get id(): string {
+        return PassphraseInteractor.ID;
+    }
+
+    async onData(data: string, cancelToken?: vscode.CancellationToken): Promise<IInteraction> {
+        const result: IInteraction = { postAction: 'keep' };
+        const lines: string[] = data.trim().split('\n');
+        if (lines.some(l => l.indexOf('Enter passphrase for') >= 0)) {
+            result.postAction = 'consume';
+            const passphrase: string | undefined = await this.passphraseProvider(undefined, undefined, cancelToken); // TODO keep track of the key name
+            if (typeof passphrase === 'string') {
+                result.response = passphrase;
+                result.isPassword = true;
+            } else {
+                result.canceled = true;
+            }
+        } else if (lines.some(l => l.indexOf('Identity added:') >= 0)) {
+            result.postAction = 'consume';
+        }
+
+        return result;
+    }
+}
+
+export function getExitCode(output: string, marker: string): number | undefined {
+    const regex: RegExp = new RegExp(`${marker}##([0-9]*)##`);
+    const match: RegExpExecArray | null = regex.exec(output);
+    if (match) {
+        try {
+            const num: number = parseInt(match[1]);
+            return Number.isNaN(num) ? undefined : num;
+        } catch (err) {
+            return undefined;
+        }
+    }
+
+    return 0;
+}
+
+/**
```

Comment:
Can you expand on this function's comment?

Reply:
What would you like to see in this comment?

Reply:
Can you please reword the comment for the function to capture what its doing?
The current comment is a bit hard to follow. I had to read the code-flow first before understanding this comment. 
```
/* 
* 's password:
 * or
 * Password:
 * not
 * 's old password:
 * 's new password:
 */
```